src/v/storage/disk_log_impl.cc

```diff
@@ -527,6 +533,22 @@ ss::future<compaction_result> disk_log_impl::compact_adjacent_segments(
     std::vector<ss::lw_shared_ptr<segment>> segments;
     std::copy(range.first, range.second, std::back_inserter(segments));
 
+    bool all_segments_self_compacted = std::all_of(
```

Comment:
Note for reviewers, we already check for this condition in do_compact()

```
// loop until we compact segment or reached end of segments set
    while (!_segs.empty()) {
        const auto end_it = std::prev(_segs.end());
        auto seg_it = std::find_if(
          _segs.begin(),
          end_it,
          [offsets_compactible](ss::lw_shared_ptr<segment>& s) {
              return !s->has_appender() && s->is_compacted_segment()
                     && !s->finished_self_compaction(). <===
                     && offsets_compactible(*s);
          });
        // nothing to compact
        if (seg_it == end_it) {
            break;
        }
```

but that is not tight enough as segments may be appended to _segs after the loop breaks and non self compacted segment could be appended to _segs which is then considered for adjacent segment compaction. This is the reason why the bug was exposed in when rolling in a tight loop along with compaction tests.